Utils/Elasticsearch/Oracle2JSON.py

```diff
@@ -20,44 +22,141 @@ def connectDEFT_DSN(dsn):
+        sys.stderr.write('Could not read config file %s' % conf)
 
-    global dsn
-    dsn = Config.get("oracle", "dsn")
     conn, cursor = connectDEFT_DSN(dsn)
+    process(conn, offset_date, final_date, step_hours, queries)
+
+
+def plain(conn, queries, offset_date, end_date):
+    tasks = query_executor(conn, queries['tasks']['file'], offset_date, end_date)
+    for task in tasks:
+        task['phys_category'] = get_category(task)
+        # send NDJSON string to STDOUT
+        sys.stdout.write(json.dumps(task) + '\n')
+
+
+def squash(conn, queries, offset_date, end_date):
+    tasks = query_executor(conn, queries['tasks']['file'], offset_date, end_date)
+    datasets = query_executor(conn, queries['datasets']['file'], offset_date, end_date)
+    join_results(tasks, squash_records(datasets))
+
+def squash_records(rec):
+    """
+    a single-pass iterator (for a generator) restructuring list of datasets: 
+    from view: 
+    [
+       {"taskid": 1, "type": "input", "datasetname": "First_in"},
+       {"taskid": 1, "type": "input", "datasetname": "Second_in"},
+       {"taskid": 1, "type": "input", "datasetname": "Third_in"},
+       {"taskid": 1, "type": "output", "datasetname": "First_out"},
+       {"taskid": 1, "type": "output", "datasetname": "Second_out"},
+       {"taskid": 1, "type": "output", "datasetname": "Third_out"},
+       {"taskid": 2, "type": "input", "datasetname": "First_in"},
+       {"taskid": 2, "type": "output", "datasetname": "First_out"},
+       {"taskid": 2, "type": "output", "datasetname": "Second_out"},
+    ...]
+     to 
+    [
+       {"taskid": 1, 
+        "input": ["First_in", "Second_in", "Third_in"], 
+        "output": ["First_out", "Second_out", "Third_out"]
+       },
+       {"taskid": 2, 
+        "input": ["First_in"], 
+        "output": ["First_out","Second_out"]
+       },
+    ...]
+    """
+    grouper = defaultdict(lambda: defaultdict(list))
+    for d in rec:
+        grouper[d['taskid']][d['type']].append(d['datasetname'])
+    return [dict(taskid=k, **v) for k, v in grouper.items()]
+
+
+def join_results(tasks, datasets):
+    d = defaultdict(dict)
+    for l in (tasks, datasets):
+        for elem in l:
+            d[elem['taskid']].update(elem)
+            sys.stdout.write(json.dumps(str(d[elem['taskid']])) + '\n')
+
+def process(conn, offset_date, final_date, step_hours, queries):
+    while (datetime.strptime(offset_date, "%d-%m-%Y %H:%M:%S") < datetime.strptime(final_date, "%d-%m-%Y %H:%M:%S")):
+        # get offset from configuration file
+        offset_date = get_offset()
+        end_date = (datetime.strptime(offset_date, "%d-%m-%Y %H:%M:%S") +
+                    timedelta(hours=step_hours)).strftime("%d-%m-%Y %H:%M:%S")
+        if mode == 'SQUASH':
+            squash(conn, queries, offset_date, end_date)
+        elif mode == 'PLAIN':
+            plain(conn, queries, offset_date, end_date)
+        update_offset(end_date)
```

Comment:
Here we update offset with the `end_date`, but the end date can be later then `now` -- \`cause it is defined as `offset + step_hours`.
It means that when we start the program next time, we can miss records, that were updated between the the real time when the last request was made, and the `end_date`.